File: csharp/Betfair.ESAClient/Betfair.ESAClient/Cache/MarketCache.cs
using Betfair.ESAClient.Protocol;
using Betfair.ESASwagger.Model;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Cache
{
    /// <summary>
    /// Thread safe cache of markets
    /// </summary>
    public class MarketCache
    {
        private readonly ConcurrentDictionary<string, Market> _markets = new ConcurrentDictionary<string, Market>();
        
        public MarketCache()
        {
            IsMarketRemovedOnClose = true;
        }

        /// <summary>
        /// Conflation indicates slow consumption
        /// </summary>
        public int ConflatedCount { get; internal set; }

        public void OnMarketChange(ChangeMessage<MarketChange> changeMessage)
        {
            if (changeMessage.IsStartOfNewSubscription)
            {
                //clear cache
                _markets.Clear();
            }
            if(changeMessage.Items != null)
            {
                //lazy build events
                List<MarketChangedEventArgs> batch = BatchMarketsChanged == null ? null : new List<MarketChangedEventArgs>(changeMessage.Items.Count);           

                foreach (MarketChange marketChange in changeMessage.Items)
                {
                    Market market = OnMarketChange(marketChange);

                    if(IsMarketRemovedOnClose && market.IsClosed)
                    {
                        //remove on close
                        Market removed;
                        _markets.TryRemove(market.MarketId, out removed);
                    }

                    //lazy build events
                    if(batch != null || MarketChanged != null)
                    {
                        MarketChangedEventArgs arg = new MarketChangedEventArgs() { Change = marketChange, Market = market };
                        if(MarketChanged != null)
                        {
                            DispatchMarketChanged(arg);
                        }
                        if(batch != null)
                        {
                            batch.Add(arg);
                        }
                    }
                }
                if(batch != null)
                {
                    DispatchBatchMarketsChanged(new BatchMarketsChangedEventArgs() { Changes = batch });
                }
            }
        }



        private Market OnMarketChange(MarketChange marketChange)
        {
            if(marketChange.Con == true)
            {
                ConflatedCount++;
            }
            Market market = _markets.GetOrAdd(marketChange.Id, id => new Market(this, id));
            market.OnMarketChange(marketChange);
            return market;
        }

        private void DispatchMarketChanged(MarketChangedEventArgs args)
        {
            try
            {
                MarketChanged.Invoke(this, args);
            }
            catch(Exception e)
            {
                Trace.TraceError("Error dispatching event: {0}", e);
            }            
        }

        private void DispatchBatchMarketsChanged(BatchMarketsChangedEventArgs args)
        {
            try
            {
                BatchMarketsChanged.Invoke(this, args);
            }
            catch (Exception e)
            {
                Trace.TraceError("Error dispatching event: {0}", e);
            }
        }

        /// <summary>
        /// Wether markets are automatically removed on close
        /// (default is true)
        /// </summary>
        public bool IsMarketRemovedOnClose { get; set; }

        /// <summary>
        /// Event for each market change
        /// </summary>
        public event MarketChangedEventHandler MarketChanged;

        /// <summary>
        /// Event for each batch of market changes
        /// (note to be truly atomic you will want to set to merge segments
        /// otherwise an event could be segmented)
        /// </summary>
        public event BatchMarketsChangedEventHandler BatchMarketsChanged;

        /// <summary>
        /// Queries by market id - the result is invariant for the 
        /// lifetime of the market.
        /// </summary>
        /// <param name="marketid"></param>
        /// <returns></returns>
        public Market this[string marketid]
        {
            get
            {
                return _markets[marketid];
            }
        }

        /// <summary>
        /// All the cached markets
        /// </summary>
        public IEnumerable<Market> Markets
        {
            get
            {
                return _markets.Values;
            }
        }

        /// <summary>
        /// Market count
        /// </summary>
        public int Count
        {
            get
            {
                return _markets.Count;
            }
        }
    }

    public delegate void MarketChangedEventHandler(object sender, MarketChangedEventArgs e);

    public delegate void BatchMarketsChangedEventHandler(object sender, BatchMarketsChangedEventArgs e);

    public class MarketChangedEventArgs : EventArgs
    {
        /// <summary>
        /// The raw change message that was just applied
        /// </summary>
        public MarketChange Change { get; internal set; }

        /// <summary>
        /// The market changed - this is reference invariant
        /// </summary>
        public Market Market { get; internal set; }

        /// <summary>
        /// Takes or returns an existing immutable snap of the market.
        /// </summary>
        public MarketSnap Snap
        {
            get
            {
                return Market.Snap;
            }
        }
    }

    public class BatchMarketsChangedEventArgs : EventArgs
    {
        public IList<MarketChangedEventArgs> Changes { get; internal set; }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/cache/market/MarketCache.java
package com.betfair.esa.client.cache.market;

import com.betfair.esa.client.protocol.ChangeMessage;
import com.betfair.esa.swagger.model.MarketChange;
import java.util.ArrayList;
import java.util.EventObject;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Thread safe cache of markets */
public class MarketCache {
    private static final Logger LOG = LoggerFactory.getLogger(MarketCache.class);

    private final Map<String, Market> markets = new ConcurrentHashMap<>();
    // whether markets are automatically removed on close (default is True)
    private boolean isMarketRemovedOnClose;
    // conflation indicates slow consumption
    private int conflatedCount;

    private final CopyOnWriteArrayList<MarketChangeListener> marketChangeListeners =
            new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<BatchMarketsChangeListener> batchMarketChangeListeners =
            new CopyOnWriteArrayList<>();

    public MarketCache() {
        this.isMarketRemovedOnClose = true;
    }

    public void onMarketChange(ChangeMessage<MarketChange> changeMessage) {
        if (changeMessage.isStartOfNewSubscription()) {
            // clear cache
            markets.clear();
        }
        if (changeMessage.getItems() != null) {
            // lazy build events
            List<MarketChangeEvent> batch =
                    (batchMarketChangeListeners.size() == 0)
                            ? null
                            : new ArrayList<>(changeMessage.getItems().size());

            for (MarketChange marketChange : changeMessage.getItems()) {
                Market market = onMarketChange(marketChange);

                if (isMarketRemovedOnClose && market.isClosed()) {
                    // remove on close
                    markets.remove(market.getMarketId());
                }

                // lazy build events
                if (batch != null || marketChangeListeners.size() != 0) {
                    MarketChangeEvent marketChangeEvent = new MarketChangeEvent(this);
                    marketChangeEvent.setChange(marketChange);
                    marketChangeEvent.setMarket(market);
                    if (marketChangeListeners != null) {
                        dispatchMarketChanged(marketChangeEvent);
                    }
                    if (batch != null) {
                        batch.add(marketChangeEvent);
                    }
                }
            }
            if (batch != null) {
                dispatchBatchMarketChanged(new BatchMarketChangeEvent(batch));
            }
        }
    }

    private Market onMarketChange(MarketChange marketChange) {
        if (Boolean.TRUE.equals(marketChange.isCon())) {
            conflatedCount++;
        }
        Market market = markets.computeIfAbsent(marketChange.getId(), k -> new Market(this, k));
        market.onMarketChange(marketChange);
        return market;
    }

    public int getConflatedCount() {
        return conflatedCount;
    }

    void setConflatedCount(int conflatedCount) {
        this.conflatedCount = conflatedCount;
    }

    public boolean isMarketRemovedOnClose() {
        return isMarketRemovedOnClose;
    }

    public void setMarketRemovedOnClose(boolean marketRemovedOnClose) {
        isMarketRemovedOnClose = marketRemovedOnClose;
    }

    public Market getMarket(String marketId) {
        // queries by market id - the result is invariant for the lifetime of the market.
        return markets.get(marketId);
    }

    public Iterable<Market> getMarkets() {
        // all the cached markets
        return markets.values();
    }

    public int getCount() {
        // market count
        return markets.size();
    }

    // Event for each market change

    private void dispatchMarketChanged(MarketChangeEvent marketChangeEvent) {
        try {
            marketChangeListeners.forEach(l -> l.marketChange(marketChangeEvent));
        } catch (Exception e) {
            LOG.error("Exception from event listener", e);
        }
    }

    public void addMarketChangeListener(MarketChangeListener marketChangeListener) {
        marketChangeListeners.add(marketChangeListener);
    }

    public void removeMarketChangeListener(MarketChangeListener marketChangeListener) {
        marketChangeListeners.remove(marketChangeListener);
    }

    // Event for each batch of market changes
    // (note to be truly atomic you will want to set to merge segments otherwise an event could be
    // segmented)

    private void dispatchBatchMarketChanged(BatchMarketChangeEvent batchMarketChangeEvent) {
        try {
            batchMarketChangeListeners.forEach(l -> l.batchMarketsChange(batchMarketChangeEvent));
        } catch (Exception e) {
            LOG.error("Exception from batch event listener", e);
        }
    }

    public void addBatchMarketChangeListener(BatchMarketsChangeListener batchMarketChangeListener) {
        batchMarketChangeListeners.add(batchMarketChangeListener);
    }

    public void removeBatchMarketChangeListener(
            BatchMarketsChangeListener batchMarketChangeListener) {
        batchMarketChangeListeners.remove(batchMarketChangeListener);
    }

    // Listeners

    public class MarketChangeEvent extends EventObject {
        // the raw change message that was just applied
        private MarketChange change;
        // the market changed - this is reference invariant
        private Market market;

        /**
         * Constructs a prototypical Event.
         *
         * @param source The object on which the Event initially occurred.
         * @throws IllegalArgumentException if source is null.
         */
        public MarketChangeEvent(Object source) {
            super(source);
        }

        public MarketChange getChange() {
            return change;
        }

        void setChange(MarketChange change) {
            this.change = change;
        }

        public Market getMarket() {
            return market;
        }

        void setMarket(Market market) {
            this.market = market;
        }

        public MarketSnap getSnap() {
            return market.getSnap();
        }
    }

    public class BatchMarketChangeEvent extends EventObject {
        private List<MarketChangeEvent> changes;

        /**
         * Constructs a prototypical Event.
         *
         * @param source The object on which the Event initially occurred.
         * @throws IllegalArgumentException if source is null.
         */
        public BatchMarketChangeEvent(Object source) {
            super(source);
        }

        public List<MarketChangeEvent> getChanges() {
            return changes;
        }

        void setChanges(List<MarketChangeEvent> changes) {
            this.changes = changes;
        }
    }

    public interface MarketChangeListener extends java.util.EventListener {
        void marketChange(MarketChangeEvent marketChangeEvent);
    }

    public interface BatchMarketsChangeListener extends java.util.EventListener {
        void batchMarketsChange(BatchMarketChangeEvent batchMarketChangeEvent);
    }
}
